### ISSUE TITLE
DivNode canonicalize phase don't remove the RemNode

### ISSUE THREAD

**Issue by qc1iu** [oracle/graal#2010]
In DivNode canonicalize phase, it can convert the expression `((a - a % b) / b)` into `(a / b)`, but it seems remains a redundant RemNode which should be eliminated.

``` java
// test case
int foo(int a, int b) {
     int r = (a - a % b) / b;
     return r;
}
```

``` asm
// generated by Graal 
0x0000ffff8ff72d10: stp         x29, x30, [sp, #32]
0x0000ffff8ff72d14: cbz         w3, 0x0000ffff8ff72d38
0x0000ffff8ff72d18: sdiv        w0, w2, w3 (redundant?)
0x0000ffff8ff72d1c: msub        w0, w0, w3, w2 (redundant?)
0x0000ffff8ff72d20: sdiv        w0, w2, w3
0x0000ffff8ff72d24: ldp         x29, x30, [sp, #32]
```

I checked GCC, go and C2(recently version),  and all of them don't have this optimization. 

**Are there any cases like `(a - a % b) / b` in real workload applications or benchmarks?** If so, if it means the RemNode needs to be eliminated.

**Comment by qc1iu** [comment]
@mur47x111 If it‘s necessary to remove the rem node in `DivOptimizationPhase`

**Comment by gergo-** [comment]
@qc1iu I will look into this when I'm back from my vacation. Thank you for your patience.

**Comment by gergo-** [comment]
I have a fix that has passed our internal tests. I hope to merge it soon.

**Comment by qc1iu** [comment]
Thanks for your attention.

And I also wondering if there are any cases like `(a - a % b) / b` in real workload applications or benchmarks? I mean... after all, it take some times for compiler to match this pattern.

**Comment by gergo-** [comment]
Thank you again for the report. This is fixed by af23840. Please test and reopen if for some reason this doesn't work for you.
I have seen this canonicalization triggered on real-world code while testing, although it is indeed very rare. However, the check is cheap. There are certainly more promising areas for optimization of compile time.
